File: 백준/Silver/9012. 괄호/괄호.py
import sys

def valid(ps):
    stk = []
    for c in ps:
        if c == '(':
            stk.append(c)
        else:
            if stk:
                stk.pop()
            else:
                return False
    
    return len(stk) == 0


n = int(sys.stdin.readline())

for _ in range(n):
    ps = sys.stdin.readline().strip()
    
    print("YES" if valid(ps) else "NO")

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>
#include <stack>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);

using namespace std;

bool valid(string ps){
    stack<char> stk;
    for(char c : ps){
        if(c == '(') stk.push(c);
        else{
            if(stk.empty()) return false;
            else stk.pop();
        }
    }
    
    return stk.empty();
}

int main()
{
    FASTIO;
    
    int n;
    cin >> n;
    
    for(int i = 0; i < n; i++){
        string ps;
        cin >> ps;
        
        cout << (valid(ps) ? "YES\n" : "NO\n");
    }

    return 0;
}
